### PR TITLE
Support CMake in Test.py

### PR DESCRIPTION
**Pull request by ximinez**
* scons will remain the default build type.
* New set of command line switches to control CMake.

...The list of functionality needing scons keeps getting shorter...

### CODE REVIEW

File: Builds/Test.py [hunk 1]
```diff
@@ -61,6 +66,41 @@ def powerset(iterable):
...
+                   for unity, targets in CMAKE_UNITY_COMBOS.iteritems() }
+else:
+    CMAKE_DIR_TARGETS = { (cc + "." + build + unity,) : targets
+                   for cc in ['gcc', 'clang']
+                   for build in ['debug', 'release', 'coverage', 'profile']
+                   for unity, targets in CMAKE_UNITY_COMBOS.iteritems() }
+
+# list of tuples of all possible options
+if IS_WINDOWS or IS_OS_X:
+    CMAKE_ALL_GENERATE_OPTIONS = [tuple(x) for x in powerset(['-Dassert=true'])]
```
Comment (seelabs) [inline Builds/Test.py]: Using powerset here is a bit overkill. Just a literal `= [(), ('-Dassert=true')]` should work too. I see why you did it though; it echos other code. Your call if you want to keep it as is.

File: Builds/CMake/CMakeFuncs.cmake [hunk 2]
```diff
@@ -35,6 +35,7 @@ macro(parse_target)
       string(REGEX REPLACE "^\\.?[^\\.]+(.*$)" "\\1" remaining ${remaining})
 
       if (${cur_component} STREQUAL gcc)
+        set(expect_cc true)
```
Comment (seelabs) [inline Builds/CMake/CMakeFuncs.cmake]: Q: I'm trying to figure out scenario `expect_cc` protects against. If we didn't have this, and unconditionally check if it's set (and not `-NOTFOUND`) before we set the cache, what bad thing happens?

File: CMakeLists.txt [hunk 3]
```diff
@@ -22,17 +22,25 @@
...
-#    a 32-bit project will be created. There is no way to set the generator or
-#    force a 64-bit build in CMakeLists.txt (setting CMAKE_GENERATOR_PLATFORM won't work).
-#    The best solution may be to wrap cmake with a script.
+#  * Use the -G"Visual Studio 14 2015 Win64" generator, or the "VS2015 x86 x64
+#    Cross Tools" Command Prompt on Windows. Without this a 32-bit project will be
+#    created. There is no way to set the generator or force a 64-bit build in
+#    CMakeLists.txt (setting CMAKE_GENERATOR_PLATFORM won't work). The best solution
+#    may be to wrap cmake with a script.
+#
+#  * Ninja commmand line builds seem to work under Windows, but only from within
```
Comment (seelabs) [inline CMakeLists.txt]: `commmand` -> `command`

File: Builds/Test.py (hunk 1 above)
Comment (ximinez) [inline Builds/Test.py]: I've got a function named `powerset`, of COURSE it's overkill! 😉 
I would like to keep it that way to express intent, as well as allow potential future expansion if more options get supported in Windows / OSX builds in the future.

File: Builds/CMake/CMakeFuncs.cmake (hunk 2 above)
Comment (ximinez) [inline Builds/CMake/CMakeFuncs.cmake]: A: It was an edge case setting up Ninja builds under Windows.
https://github.com/ximinez/rippled/blob/4e693d611ef8965114f3a5689e9f9f8b4476187f/CMakeLists.txt#L31-L32
> * Ninja command line builds seem to work under Windows, but only from within
>    the "VS2015 x86 x64 Cross Tools" Command Prompt.
Without this change, if I'm under that command prompt, and run `cmake ../../.. -GNinja`, the `project` command complains that it can't find a compiler - because this block wrote out an empty string.

I'll add a brief comment.

File: CMakeLists.txt (hunk 3 above)
Comment (ximinez) [inline CMakeLists.txt]: Fixed.

File: Builds/CMake/CMakeFuncs.cmake [hunk 4]
```diff
@@ -610,11 +624,11 @@ macro(setup_build_boilerplate)
...
       /W3                  # Warning level 3
       /WX-                 # Disable warnings as errors
-      /wd"4018"
-      /wd"4244"
-      /wd"4267"
-      /wd"4800"            # Disable C4800(int to bool performance)
-      /wd"4503"            # Decorated name length exceeded, name was truncated
+      /wd4018
+      /wd4244
+      /wd4267
```
Comment (MarkusTeufelberger) [inline Builds/CMake/CMakeFuncs.cmake]: Might be nice to document what these are disabling and why

File: Builds/CMake/CMakeFuncs.cmake (hunk 2 above)
Comment (seelabs) [inline Builds/CMake/CMakeFuncs.cmake]: If we could check `CMAKE_C_COMPILER` and `CMAKE_CXX_COMPILER` for empty strings (or `-NOTFOUND`) before setting them in the cache, I'd rather do that. If that won't work then fine as is.

**Comment by seelabs** [comment]
:+1:, though I left one more comment on `expect_cc` for you to consider.

File: Builds/CMake/CMakeFuncs.cmake (hunk 2 above)
Comment (ximinez) [inline Builds/CMake/CMakeFuncs.cmake]: So much simpler!

File: Builds/CMake/CMakeFuncs.cmake (hunk 4 above)
Comment (ximinez) [inline Builds/CMake/CMakeFuncs.cmake]: These look like they were copied over from `SConstruct`, and they aren't documented there either, so I'll look into this, though I may need to make the change later.

**Comment by MarkusTeufelberger** [comment]
Please add a call to this to Builds/build_all.sh

It should use all available CPUs but otherwise be as vanilla as possible. I use that script to verify that rippled builds in all possible configurations on various Linux distributions. Would be nice to also run cmake builds in there.

I am not so sure if I really trust that "cmake --clean-first" will always (ninja builds...) completely clean up and would prefer if the build directory actually gets deleted.

Also you are first running all builds and only then running tests instead of running build1, test1, clean, build2, test2, clean etc. This will waste a lot of disk space (there are a LOT of different combinations possible already). It is also problematic CPU time wise, since a failing unit test will be only detected at a much later time, after everything compiled.

File: Builds/Test.py [hunk 5]
```diff
@@ -61,6 +66,41 @@ def powerset(iterable):
 IS_WINDOWS = platform.system().lower() == 'windows'
 IS_OS_X = platform.system().lower() == 'darwin'
 
+# CMake
+
+CMAKE_UNITY_TARGETS = ['rippled', 'rippled_classic']
+CMAKE_NONUNITY_TARGETS = ['rippled', 'rippled_unity']
```
Comment (MarkusTeufelberger) [inline Builds/Test.py]: rippled_unity is a nonunity target?

File: Builds/Test.py [hunk 6]
```diff
@@ -61,6 +66,41 @@ def powerset(iterable):
 IS_WINDOWS = platform.system().lower() == 'windows'
 IS_OS_X = platform.system().lower() == 'darwin'
 
+# CMake
+
```
Comment (MarkusTeufelberger) [inline Builds/Test.py]: empty line not necessary

File: Builds/Test.py [hunk 7]
```diff
@@ -61,6 +66,41 @@ def powerset(iterable):
...
+CMAKE_UNITY_TARGETS = ['rippled', 'rippled_classic']
+CMAKE_NONUNITY_TARGETS = ['rippled', 'rippled_unity']
+if IS_WINDOWS:
+    CMAKE_UNITY_CONFIGS = ['Debug', 'Release']
+    CMAKE_NONUNITY_CONFIGS = ['DebugClassic', 'ReleaseClassic']
+else:
+    CMAKE_UNITY_CONFIGS = []
+    CMAKE_NONUNITY_CONFIGS = []
+CMAKE_UNITY_COMBOS = { '' : [CMAKE_UNITY_TARGETS, CMAKE_UNITY_CONFIGS],
+    '.nounity' : [CMAKE_NONUNITY_TARGETS, CMAKE_NONUNITY_CONFIGS] }
```
Comment (MarkusTeufelberger) [inline Builds/Test.py]: I decided not to do this when I wrote the ALL_TARGETS list comprehension below, because I found it harder to read and understand the code if written like this for very little benefit (new targets rarely get added).

Please consider also simply explicitly listing the combinations instead of combining several lists with very few elements.

File: Builds/Test.py [hunk 8]
```diff
@@ -61,6 +66,41 @@ def powerset(iterable):
...
+        CMAKE_UNITY_COMBOS.iteritems() }
+elif IS_OS_X:
+    CMAKE_DIR_TARGETS = { (build + unity,) : targets
+                   for build in ['debug', 'release']
+                   for unity, targets in CMAKE_UNITY_COMBOS.iteritems() }
+else:
+    CMAKE_DIR_TARGETS = { (cc + "." + build + unity,) : targets
+                   for cc in ['gcc', 'clang']
+                   for build in ['debug', 'release', 'coverage', 'profile']
+                   for unity, targets in CMAKE_UNITY_COMBOS.iteritems() }
```
Comment (MarkusTeufelberger) [inline Builds/Test.py]: How does this work with --all? As far as I understand, several builds will end up in the same folder and only the last version will survive and get (unit-)tested.

File: Builds/Test.py [hunk 9]
```diff
@@ -129,14 +169,76 @@ def powerset(iterable):
...
+pgroup.add_argument(
+    '--cmake', '-m',
+    action='store_true',
+    help='Build using CMake.',
+)
+
+pgroup.add_argument(
+    '--scons',
+    action='store_true',
+    help='Build using Scons. Default behavior.')
```
Comment (MarkusTeufelberger) [inline Builds/Test.py]: This default behavior is implicit, not explicit (`if not ARGS.cmake`)

File: Builds/Test.py [hunk 10]
```diff
@@ -129,14 +169,76 @@ def powerset(iterable):
...
+
 parser.add_argument(
-    'scons_args',
+    '--dir', '-d',
     default=(),
-    nargs='*'
+    nargs='*',
+    help='Specify one or more CMake dir names. Implies --cmake. '
```
Comment (MarkusTeufelberger) [inline Builds/Test.py]: "Implies --cmake" - please either enforce this or prefix the switch with cmake.

File: Builds/Test.py [hunk 11]
```diff
@@ -129,14 +169,76 @@ def powerset(iterable):
     help='Reduce output where possible (unit tests)',
 )
 
+# Scons and CMake parameters are too different to run
+# both side-by-side
+pgroup = parser.add_mutually_exclusive_group()
+
+pgroup.add_argument(
+    '--cmake', '-m',
```
Comment (MarkusTeufelberger) [inline Builds/Test.py]: `-m`? This command and some others below have some quite unintuitive one letter abbreviations...

File: Builds/Test.py [hunk 12]
```diff
@@ -227,45 +327,168 @@ def run_build(args=None):
...
+    if target:
+      build_args += ('--target', target)
+    if config:
+      build_args += ('--config', config)
+    if clean:
+      build_args += ('--clean-first')
+    if args:
+        build_args += ('--',)
+        build_args += tuple(args)
+    resultcode, lines = shell('cmake', build_args)
```
Comment (MarkusTeufelberger) [inline Builds/Test.py]: I'm not a cmake expert, but I thought one needs to run `cmake [args]` first, then run `make` or `ninja` because cmake just generates a `Makefile`? If one just needs to run cmake twice, this is of course correct.

File: Builds/CMake/CMakeFuncs.cmake [hunk 13]
```diff
@@ -594,27 +608,27 @@ macro(setup_build_boilerplate)
...
-      /W3                  # Warning level 3
-      /WX-                 # Disable warnings as errors
-      /wd"4018"
-      /wd"4244"
-      /wd"4267"
-      /wd"4800"            # Disable C4800(int to bool performance)
-      /wd"4503"            # Decorated name length exceeded, name was truncated
+      /MP                # Multiprocessor compilation
+      /openmp-           # pragma omp: disabled
+      /Zc:forScope       # Language extension: for scope
```
Comment (seelabs) [inline Builds/CMake/CMakeFuncs.cmake]: Nit: `Language extension: for scope` makes it sounds like this option is enabling a language extension instead of forcing language _conformance_. Reword to `Language conformance: for scope`? 

Also, according to this: https://msdn.microsoft.com/en-us/library/84wcsx8x.aspx, this option is on by default since VS 2015, so we may not need this anymore. (I would _not_ remove the option in this changeset; cleaning up the VS options should be it's own PR).

File: Builds/Test.py [hunk 14]
```diff
@@ -227,45 +327,168 @@ def run_build(args=None):
...
+
+                    for target in build_targets:
+                        for config in build_configs:
+                            run_cmake_build(cmake_full_dir, target, config, ARGS.clean, ARGS.build_option)
+
+                for cmake_dir in build_dirs:
+                    cmake_full_dir = get_cmake_dir(cmake_dir)
+                    for target in build_targets:
+                        for config in build_configs:
+                            failed = run_cmake_tests(cmake_full_dir, target, config)
```
Comment (MarkusTeufelberger) [inline Builds/Test.py]: This should be moved directly after run_cmake_build

File: Builds/Test.py [hunk 15]
```diff
@@ -227,45 +327,168 @@ def run_build(args=None):
...
+                                else:
+                                    all_failed.extend([decodeString(cmake_dir +
+                                            "." + target + "." + config), ':'.join(f)]
+                                        for f in failed)
+                            else:
+                                print('Success')
 
     if all_failed:
-        if len(to_build) > 1:
+        if len(all_failed) > 1:
```
Comment (MarkusTeufelberger) [inline Builds/Test.py]: Not sure what I was thinking there, probably should have written a comment. I'm not sure if the length check makes sense that way though.

File: Builds/CMake/CMakeFuncs.cmake (hunk 13 above)
Comment (ximinez) [inline Builds/CMake/CMakeFuncs.cmake]: Agreed. I'll update the comment, but I'm not touching VS options in this PR.

**Comment by ximinez** [comment]
> Please add a call to this to Builds/build_all.sh

Will do.

> I am not so sure if I really trust that "cmake --clean-first" will always (ninja builds...) completely clean up and would prefer if the build directory actually gets deleted.

That's a fair point. And I discovered a neat edge case that makes `--clean-first` backfire in some circumstances, so I'll change it to delete the build directory entirely.

> Also you are first running all builds and only then running tests instead of running build1, test1, clean, build2, test2, clean etc. This will waste a lot of disk space (there are a LOT of different combinations possible already). It is also problematic CPU time wise, since a failing unit test will be only detected at a much later time, after everything compiled.

Yeah, it seems silly now, but I was basing it off of the scons-side behavior to build first, then run the tests. I'll change it.

File: Builds/Test.py (hunk 5 above)
Comment (ximinez) [inline Builds/Test.py]: This has to do with a decision we made for the cmake config specifically to support both unity and non-unity builds in a single solution in the Visual Studio IDE. If you target a unity build, `rippled` will be a unity build, and as a convenience, `rippled_classic` will be available as a nonunity build in the solution. Conversely, if you target a non-unity build, `rippled` will be a non-unity build, and `rippled_unity` will be the convenience.

File: Builds/Test.py (hunk 7 above)
Comment (ximinez) [inline Builds/Test.py]: The `*_CONFIGS` are conditional, but I'm fine with doing that for the `*_TARGETS`.

File: Builds/Test.py (hunk 8 above)
Comment (ximinez) [inline Builds/Test.py]: Even before the latest changes, the builds will end up in the same folder, but tests will be run in between each set of runs. The latest changes guarantee it.

File: Builds/Test.py (hunk 11 above)
Comment (ximinez) [inline Builds/Test.py]: Yes, unfortunately `-c` was already taken (`--clean`). If you have better suggestions for the single letter options, I'm happy to hear them.

File: Builds/Test.py (hunk 9 above)
Comment (ximinez) [inline Builds/Test.py]: Right. Eventually, cmake will be the default, so I'm putting this in place now for anybody who cares enough to stick with scons when it changes.

File: Builds/Test.py (hunk 10 above)
Comment (ximinez) [inline Builds/Test.py]: Last I checked, it is enforced. Do you have a command line where it isn't?

File: Builds/Test.py (hunk 12 above)
Comment (ximinez) [inline Builds/Test.py]: Note the first line of the function add `--build` to the command line options. CMake build tool mode abstracts away the call to `make` or `ninja` or `msbuild` so you don't have to know or care what is running over the hood, and makes it a lot easier to do stuff like this cross platform.
https://cmake.org/cmake/help/v3.6/manual/cmake.1.html#build-tool-mode
So you don't _need_ to, but running twice, the second time with `--build`, is easy and doable.

File: Builds/Test.py (hunk 14 above)
Comment (ximinez) [inline Builds/Test.py]: Agreed. Done.

File: Builds/Test.py (hunk 15 above)
Comment (ximinez) [inline Builds/Test.py]: There's an edge case if you use `--keep_going` with command line options that only do one build and the tests fail, you'll basically get this error twice. So it makes sense to only show if there was more than one build.

File: Builds/Test.py (hunk 11 above)
Comment (MarkusTeufelberger) [inline Builds/Test.py]: Maybe not all/most options need single-letter abbreviations.

File: Builds/Test.py [hunk 16]
```diff
@@ -129,14 +166,76 @@ def powerset(iterable):
...
+    nargs='*',
+    help='Specify one or more CMake dir names. Implies --cmake. '
+        'Will also be used as -Dtarget=<dir> running cmake.'
+)
+
+parser.add_argument(
+    '--target', '-r',
+    default=(),
+    nargs='*',
+    help='Specify one or more CMake build targets. Implies --cmake. '
```
Comment (mellery451) [inline Builds/Test.py]: how do I specify more than one target...say for example I want to build both `rippled` and `rippled_classic` with `Release` configuration ?

File: Builds/Test.py (hunk 11 above)
Comment (ximinez) [inline Builds/Test.py]: That's a good point. It seems easier to do without them now and add them in later than to shoehorn them in now and need to remove them later.

File: Builds/Test.py [hunk 17]
```diff
@@ -61,6 +66,38 @@ def powerset(iterable):
...
+                   for cc in ['gcc', 'clang']
+                   for build in ['debug', 'release', 'coverage', 'profile']
+                   for unity, targets in CMAKE_UNITY_COMBOS.iteritems() }
+
+# list of tuples of all possible options
+if IS_WINDOWS or IS_OS_X:
+    CMAKE_ALL_GENERATE_OPTIONS = [tuple(x) for x in powerset(['-Dassert=true'])]
+else:
+    CMAKE_ALL_GENERATE_OPTIONS = list(set(
+        [tuple(x) for x in powerset(['-GNinja', '-Dstatic=true', '-Dassert=true', '-Dsan=address'])] +
```
Comment (ximinez) [inline Builds/Test.py]: OTOH, I don't think many users actually use `--all`, so I think it's a good idea to include `-GNinja` on Windows on the assumption that if the user requests it, they should know what they're doing.

Comment (mellery451) [inline Builds/Test.py]: I'd include for all platforms in that case - `ninja` is easy to install with homebrew on mac

File: Builds/Test.py (hunk 16 above)
Comment (ximinez) [inline Builds/Test.py]: `./Builds/Test.py --target rippled rippled_classic -- -m` should do it.

File: Builds/Test.py (hunk 17 above)
Comment (MarkusTeufelberger) [inline Builds/Test.py]: Yeah, `--all` should really mean "everything in every configuration that can be built".

File: Builds/Test.py [hunk 18]
```diff
@@ -227,45 +324,163 @@ def run_build(args=None):
...
+                                print('FAILED:', *(':'.join(f) for f in failed))
+                                if not ARGS.keep_going:
+                                    sys.exit(1)
+                                else:
+                                    all_failed.extend([decodeString(cmake_dir +
+                                            "." + target + "." + config), ':'.join(f)]
+                                        for f in failed)
+                            else:
+                                print('Success')
+                    if ARGS.clean:
```
Comment (MarkusTeufelberger) [inline Builds/Test.py]: Is it enough to clean after all targets were built instead of cleaning after every single build/test run?

File: Builds/Test.py (hunk 16 above)
Comment (ximinez) [inline Builds/Test.py]: Clarification: The `-m` is a Windows build flag. On Linux, you'd want `-j<number of cores>` or similar.

File: Builds/Test.py [hunk 19]
```diff
@@ -227,45 +324,163 @@ def run_build(args=None):
...
+def run_cmake_build(directory, target, config, args):
+    print('Building', target, config, 'in', directory, 'with', *args or ('default options',))
+    build_args=('--build', directory)
+    if target:
+      build_args += ('--target', target)
+    if config:
+      build_args += ('--config', config)
+    if args:
+        build_args += ('--',)
+        build_args += tuple(args)
```
Comment (ximinez) [inline Builds/Test.py]: Cmake build mode does not abstract verbosity away, so we'd have to determine the correct flag for every build type (`msbuild` uses something like `-verbosity:d`, `make` uses `-d`, `ninja` uses `-v`, etc), and make sure we had the correct build type (which is sometimes implied, and not explicitly on the command line like `-GNinja`). I'm not sure it's worth it...

File: Builds/Test.py (hunk 18 above)
Comment (ximinez) [inline Builds/Test.py]: It should be. For each set of targets and configs, each build is done completely separately, and AFAIK doesn't do anything to change cmake config outputs, and running `run_cmake` once per set will save time.

**Comment by ximinez** [comment]
Rebased on 0.80.0-b1, squashed, and marked as Passed. I'm confident that the rebase didn't do anything nasty, but if anyone spots problems, please let me know.

**Comment by seelabs** [comment]
In 0.80.0-b2
